javaparser-core/src/main/java/com/github/javaparser/ast/nodeTypes/NodeWithExtends.java

```diff
@@ -0,0 +1,37 @@
+package com.github.javaparser.ast.nodeTypes;
+
+import java.util.List;
+
+import com.github.javaparser.ast.Node;
+import com.github.javaparser.ast.type.ClassOrInterfaceType;
+
+public interface NodeWithExtends<T> {
+    public List<ClassOrInterfaceType> getExtends();
+
+    public T setExtends(final List<ClassOrInterfaceType> extendsList);
+
+    /**
+     * Add an extends to this and automatically add the import
+     * 
+     * @param clazz the class to extand from
+     * @return this
+     */
+    public default T addExtends(Class<?> clazz) {
+        ((Node) this).tryAddImportToParentCompilationUnit(clazz);
+        return addExtends(clazz.getSimpleName());
+    }
+
+    /**
+     * Add an extends to this
+     * 
+     * @param name the name of the type to extends from
+     * @return this
+     */
+    @SuppressWarnings("unchecked")
+    public default T addExtends(String name) {
+        ClassOrInterfaceType classOrInterfaceType = new ClassOrInterfaceType(name);
+        getExtends().add(classOrInterfaceType);
+        classOrInterfaceType.setParentNode((Node) this);
+        return (T) this;

```

Comment:
Thoses interfaces are for internal use only.
They were created to avoid code duplication by taking advantage of default
methods.
There is no point of using them outside of the library or using them on a
non-subtype of Node.
Unfortunately package private exists for types but not library private :p

Le mer. 24 août 2016 18:10, chmuche <notifications@github.com> a écrit :

> In
> javaparser-core/src/main/java/com/github/javaparser/ast/nodeTypes/NodeWithExtends.java
> <https://github.com/javaparser/javaparser/pull/435#discussion_r76087938>:
>
> > +        ((Node) this).tryAddImportToParentCompilationUnit(clazz);
> > +        return addExtends(clazz.
>
> getSimpleName());
> > +    }
> > +
> > +    /**
> > +     * Add an extends to this
> > +     *
> > +     * @param name the name of the type to extends from
> > +     * @return this
> > +     */
> > +    @SuppressWarnings("unchecked")
> > +    public default T addExtends(String name) {
> > +        ClassOrInterfaceType classOrInterfaceType = new ClassOrInterfaceType(name);
> > +        getExtends().add(classOrInterfaceType);
> > +        classOrInterfaceType.setParentNode((Node) this);
> > +        return (T) this;
>
> return (T) this; will work only if this is the same type of T, your
> interface do not impose this.
> and (Node) this will work only if this extend Node or you can't be sure.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/javaparser/javaparser/pull/435/files/0f94bf05d866127fee91525aca489df4106e1b4e#r76087938>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJu56BSTWK1g_zupRYNjqX8Z0rSuAfBBks5qjGzxgaJpZM4Jrvwb>
> .
>